src/plugins/intel_cpu/src/nodes/tensoriterator.cpp

```diff
@@ -141,12 +147,134 @@ class PortIteratorHelper : public PortMapHelper {
+        mem_holder_src = from->GetPrimitive();
+        mem_holder_dst = to->GetPrimitive();
+        //if (mem_holder_src.get_desc() != mem_holder_dst.get_desc()) {
+            reorder = {mem_holder_src, mem_holder_dst};
+        //} else {
+        //    mem_holder_dst.set_data_handle(mem_holder_src.get_data_handle());
+        //}
+    }
+
+    void execute(dnnl::stream strm, int iter = -1) override {
+        if (iter != 0 && reorder.get(true)) {
+            reorder.execute(strm, mem_holder_src, mem_holder_dst);
+        }
+    }
+
+private:
+    dnnl::reorder reorder;
+    dnnl::memory mem_holder_src;
+    dnnl::memory mem_holder_dst;
+};
+
 class BackEdgePortHelper : public PortMapHelper {
 public:
-    BackEdgePortHelper(const MemoryPtr &from, const MemoryPtr &to, const dnnl::engine& eng) {
+    BackEdgePortHelper(const Node* from, Node* to) : m_from(from), m_to(to) {
+        auto from_desc = m_from->getBaseMemDescAtInputPort(inputNodePortIdx);
+        auto to_desc = m_to->getBaseMemDescAtOutputPort(outputNodePortIdx);
+        VectorDims commonDims;
+        if (from_desc->isDefined()) {
+            commonDims = from_desc->getShape().getStaticDims();
+        } else if (to_desc->isDefined()) {
+            commonDims = to_desc->getShape().getStaticDims();
+        } else {
+            auto& dimsFromMax = from_desc->getShape().getMaxDims();
+            auto& dimsToMax = to_desc->getShape().getMaxDims();
+            VectorDims commonMax(dimsFromMax.size());
+            for (size_t i = 0; i < dimsFromMax.size(); ++i) {
+                commonMax[i] = std::min(dimsFromMax[i], dimsToMax[i]);
+            }
+
+            auto& dimsFromMin = from_desc->getShape().getMinDims();
+            auto& dimsToMin = to_desc->getShape().getMinDims();
+            VectorDims commonMin(dimsFromMin.size());
+            for (size_t i = 0; i < dimsFromMax.size(); ++i) {
+                commonMin[i] = std::max(dimsFromMin[i], dimsToMin[i]);
+            }
+
+            Shape commonShape(commonMin, commonMax);
+            commonDims = MemoryDescUtils::makeDummyShape(commonShape).getStaticDims();
+        }
+
+        if (!from_desc->isDefined()) {
+            from_desc = from_desc->cloneWithNewDims(commonDims);
+        }
+
+        if (!to_desc->isDefined()) {
+            to_desc = to_desc->cloneWithNewDims(commonDims);
+        }
+
+        m_compatible = to_desc->isCompatible(*from_desc);
+        m_dynamic = m_from->isDynamicNode() || m_to->isDynamicNode();
+
+        auto from_mem_ptr = m_from->getParentEdgeAt(inputNodePortIdx)->getMemoryPtr();
+
+        if (m_compatible) {
+            auto memMngPtr = from_mem_ptr->getDnnlMemoryMngr();
+            auto& childEdges = m_to->getChildEdgesAtPort(outputNodePortIdx);
+            for (size_t i = 0; i < childEdges.size(); ++i) {
+                auto to_mem_ptr = childEdges[i]->getMemoryPtr();
+
+                to_mem_ptr->Create(to_mem_ptr->getDesc(), memMngPtr);
```

Comment:
We cannot make m_to reuse m_from's memory before the first iteration.

Think about this situation : there are more than one m_to that are back edged from the same m_from. It is quite the case of DIEN.

For the first iteration, each m_to might get the different input from outside graph respectively.
If they are sharing the same memory of m_from, the first iteration would be wrong.

So we have to delay the memory reuse at least after the first iteration.